contracts/validator/ValidatorSet.sol

```diff
@@ -50,4 +50,80 @@ contract ValidatorSet {
      *   - less than MAX_FUTURE_END_HEIGHT: for logged out validators.
      */
     mapping(address => uint256) public validatorEndHeight;
+
+
+    /* Internal functions  */
+
+    /**
+     * @notice It is for inserting validators into the validator set.
+     *
+     * @dev Function requires :
+     *          - Validator address must not be 0.
+     *          - Validator address is already used.
+     *
+     * @param _validator Validator address.
+     * @param _beginHeight Begin height for the validator.
+     */
+    function insertValidator(
+        address _validator,
+        uint256 _beginHeight
+    )
+        internal
+    {
+        require(
+            _validator != address(0),
+            "Validator must not be null address."
+        );
+        require(
```

Comment:
I believe, there was an agreement that validator set is not going to fail if validator already exists on insert (will ignore) and also if validator is not there on removal. Yes? /cc @deepesh-kn @jayeshbairagi

Reply:
I remember that we had a discussion that `upsert` will not fail when it calls the `insertValidator` and `removeValidator` functions. It will have all the require condition in place.
Function `upsert` will check if the validators do not exist, then it will call `insertValidator`, if it exists then it will not call this function. 
Similarly, if the reputation is 0 then it will check if its already deregistered, if not then it will call `removeValidator` else not.

Reply:
Both @deepesh-kn and @pgev s points can be true at the same time; `Protocore:upsert` and `core:join/logout` can do the required checks, and `ValidatorSet` can optimise by not doing these checks, because we will have to do them double (`v!=0`).  

For Gen1 I am not concerned about the extra gas use, and let's just `assert` or `require`. An `assert` perhaps can be a compromise.

Reply:
@gulshanvasnani can you make these `asserts`?

Reply:
Changed `requires` to `assert` in insert validator method.